modules/diskq/tests/test_diskq_truncate.c

```diff
@@ -0,0 +1,172 @@
+  gint number_of_msgs_to_pop;
+  gboolean should_file_be_empty_after_truncate;
+} truncate_test_params;
+
+static void
+_test_diskq_truncate(truncate_test_params params)
+{
+  LogQueue *q;
+  DiskQueueOptions options;
+  GString *filename = g_string_new("test_dq_truncate.qf");
+
+  q = _create_diskqueue(filename->str, &options);
+  assert_gint(log_queue_get_length(q), 0, "Invalid queue length!");
+
+  feed_some_messages(q, params.number_of_msgs_to_push, &parse_options);
+  assert_gint(log_queue_get_length(q), params.number_of_msgs_to_push, "Invalid queue length!");
+  gint64 size_after_push = ((LogQueueDisk *)q)->qdisk->file_size;
+  struct stat after_push;
+  stat(filename->str, &after_push);
+  assert_gint(size_after_push, after_push.st_size, "File size does not match with stored size");
+
+  send_some_messages(q, params.number_of_msgs_to_pop);
+  assert_gint(log_queue_get_length(q),
+              params.number_of_msgs_to_push - params.number_of_msgs_to_pop,
+              "Invalid queue length1!");
+
+  _save_diskqueue(q);
+
+  q = _get_diskqueue(filename->str, &options);
+  assert_gint(log_queue_get_length(q),
+              params.number_of_msgs_to_push - params.number_of_msgs_to_pop,
+              "Invalid queue length2!");
+
+  gint64 size_after_truncate = ((LogQueueDisk *)q)->qdisk->file_size;
+  struct stat after_truncate;
+  stat(filename->str, &after_truncate);
+  fprintf(stderr, "Disk-queue file size original:%lu after truncate:%lu\n", size_after_push, size_after_truncate);
+
+  if (params.should_file_be_empty_after_truncate)
+    assert_gint(size_after_truncate, QDISK_RESERVED_SPACE, "Truncate failed: file is not empty");
+  else
+    assert_gint(size_after_truncate, after_truncate.st_size, "Truncate failed: size does not match with stored");
+
+
+  unlink(filename->str);
+  g_string_free(filename, TRUE);
+
+  log_queue_unref(q);
+  disk_queue_options_destroy(&options);
+}
+
+static void
+test_diskq_truncate_with_diskbuffer_used(void)
+{
+  _test_diskq_truncate((truncate_test_params)
+  {
+    .number_of_msgs_to_push = 100,
+     .number_of_msgs_to_pop = 50,
+      .should_file_be_empty_after_truncate = FALSE
+  });
+}
+
+static void
+test_diskq_truncate_without_diskbuffer_used(void)
+{
+  _test_diskq_truncate((truncate_test_params)
+  {
+    .number_of_msgs_to_push = 100,
+     .number_of_msgs_to_pop = 80,
+      .should_file_be_empty_after_truncate = TRUE
+  });
+}
+
+
+int
+main(void)
+{
+  msg_init(TRUE); // internal messages will go to stderr(), msg_init() is idempotent
+  app_startup();
+
+  configuration = cfg_new_snippet();
+  assert_true(cfg_init(configuration), "cfg_init failed!");
+  cfg_load_module(configuration, "syslogformat");
+  msg_format_options_defaults(&parse_options);
+  msg_format_options_init(&parse_options, configuration);
+
+  // File-queue is the part of disk-queue that is used when qout is full
+  test_diskq_truncate_with_diskbuffer_used();
+  test_diskq_truncate_without_diskbuffer_used();
+
```

Comment:
I miss deinitialization: like cfg_free, and app_shutdown.

Reply:
will fix.